File: problems/problems_1353/Solution.java
package problems.problems_1353;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    private int find(int[] fa, int x) {
        if (fa[x] != x) {
            fa[x] = find(fa, fa[x]);
        }
        return fa[x];
    }

    public int maxEvents(int[][] events) {
        Arrays.sort(events, Comparator.comparingInt(o -> o[1]));
        int maxDay = events[events.length - 1][1];
        int[] fa = new int[maxDay + 2];
        for (int i = 0; i < fa.length; ++i) {
            fa[i] = i;
        }
        int ans = 0;
        for (int[] event: events) {
            int x = find(fa, event[0]);
            if (x <= event[1]) {
                ++ans;
                fa[x] = x + 1;
            }
        }
        return ans;
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        int[][] events = jsonArrayToInt2DArray(inputJsonValues[0]);
        return JSON.toJSON(maxEvents(events));
    }
}

File: problems/problems_1353/solution.py
import heapq

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.maxEvents(test_input)

    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()
        cur_day = events[0][0]
        ans = 0
        idx = 0
        pq = []
        n = len(events)
        while idx < n or pq:
            while idx < n and events[idx][0] <= cur_day:
                heapq.heappush(pq, events[idx][1])
                idx += 1
            if pq:
                heapq.heappop(pq)
                ans += 1
                cur_day += 1
                while pq and pq[0] < cur_day:
                    heapq.heappop(pq)
            elif idx < n:
                cur_day = max(cur_day, events[idx][0])
        return ans
